python/tvm/topi/cuda/scatter.py

```diff
@@ -789,42 +790,80 @@ def gen_ir(data_ptr, indices_ptr, updates_ptr, out_ptr):
-                else:
-                    raise NotImplementedError("scatter_nd mode not in [update, add]:", mode)
+        with ib.new_scope():
+            if updates.dtype == "int64" and mode == "add":
+                bdim_x = ceil_div(fused_updates_dimension, tdim)
+                bx = te.thread_axis("blockIdx.x")
+                tx = te.thread_axis("threadIdx.x")
+                ib.scope_attr(bx, "thread_extent", bdim_x)
+                ib.scope_attr(tx, "thread_extent", tdim)
+                with ib.for_range(0, fused_indices_dimension) as i:
+                    j = bx * tdim + tx
+                    with ib.if_scope(j < fused_updates_dimension):
+                        offset = fused_updates_dimension
+                        index = j  # This is x_M, .. x_{N-1} part of the index into out.
+                        # Build up the
+                        # indices[0, y_0, .. y_{K-1}], ... indices[M-1, y_0, .. y_{K-1}]
+                        # part of the index into out.
+                        for l in reversed(range(indices_ptr.shape[0].value)):
+                            # indices[i * l * fused_indices_dimension] = indices[l, y_0,
+                            #                                                   ... y_{k-1}]
+                            index += offset * indices[i + l * fused_indices_dimension]
+                            offset *= data_ptr.shape[l]
+                        out[index] += updates[i * fused_updates_dimension + j]
+            else:
+                bdim_x = ceil_div(fused_updates_dimension, tdim)
+                bdim_y = fused_indices_dimension
+                # In case of large input sizes, bim_y might be too large.
+                # So it could be moved to blockIdx.x position, which holds larger scales.
+                bx = te.thread_axis("blockIdx.y")
+                by = te.thread_axis("blockIdx.x")
+                tx = te.thread_axis("threadIdx.x")
+                ib.scope_attr(bx, "thread_extent", bdim_x)
+                ib.scope_attr(by, "thread_extent", bdim_y)
+                ib.scope_attr(tx, "thread_extent", tdim)
+
+                j = bx * tdim + tx
+                with ib.if_scope(j < fused_updates_dimension):
+                    offset = fused_updates_dimension
+                    index = j  # This is x_M, .. x_{N-1} part of the index into out.
+                    # Build up the indices[0, y_0, .. y_{K-1}], .. indices[M-1, y_0, .. y_{K-1}]
+                    # part of the index into out.
+                    up_index = by * fused_updates_dimension + j
+                    for l in reversed(range(indices_ptr.shape[0].value)):
+                        # indices[by * l * fused_indices_dimension] = indices[l, y_0, ... y_{k-1}]
+                        index += offset * indices[by + l * fused_indices_dimension]
+                        offset *= data_ptr.shape[l]
+                    if mode == "update":
+                        out[index] = updates[up_index]
```

Comment:
For `update` mode, does this give deterministic output?

Reply:
> The output is calculated via the following equation:
> ```
>   output = np.copy(data)
>   update_indices = indices.shape[:-1]
>   for idx in np.ndindex(update_indices):
>       output[indices[idx]] = updates[idx]
> ```
> The order of iteration in the above loop is not specified. In particular, indices should not have duplicate entries: that is, if idx1 != idx2, then indices[idx1] != indices[idx2]. This ensures that the output value does not depend on the iteration order.

@masahi 
According to the [defination of ScatterND](https://github.com/onnx/onnx/blob/master/docs/Operators.md#scatternd) in ONNX, output does not depend on the iteration order. 

Based on the above assumption, we replace the original `with ib.for_range(0, fused_indices_dimension) as i` with `blockIdx.y`.

Is this what you concerned ?

Reply:
It doesn't matter what ONNX says. If the previous implementation gives a deterministic output, performance improvement shouldn't break that. If you use atomic for `add` mode, then I assume that multiple threads compete for the same write index. This leads to non-determinstic output for `update` mode.

Reply:
@masahi 
I see. So, should I fallback to previous algorithm when `mode="update"`? Or any suggestions. Thanks.

Reply:
Yes, for `update` the prev algo should be used.

Reply:
OK, I'll fallback to previous algo for `update`. 
For `allow_non_deterministic` feature, maybe we could fire a new PR?

Reply:
Yes that sounds good, we can discuss with more people then. This has been on my mind for a while, since both our `scatter` and `scatter_nd` op sacrifice performance for deterministic output, while all other frameworks make the opposite choice (they say output is undefined when indices are not unique).

Reply:
Great! Looking forward to further improvement